modules/javafx.graphics/src/test/java/test/javafx/scene/CssStyleHelperTest.java

```diff
@@ -658,4 +668,90 @@ public void initialNodeWithUserSetValueShouldNotResetValuesOnOtherNodesWithoutOv
         assertEquals(new Insets(4), b.getPadding());
     }
+
+    private static final String GRAY_STYLESHEET = ".my-pane {-fx-background-color: #808080}";
+    private static final String GRAY_INDIRECT_STYLESHEET = ".root {-fx-base: #808080} .my-pane {-fx-background-color: -fx-base}";
+    private static final String RED_STYLESHEET = ".my-pane {-fx-background-color: red}";
+    private static final String RED_INDIRECT_STYLESHEET = ".root {-fx-base: red} .my-pane {-fx-background-color: -fx-base}";
+    private static final String FX_BASE_GREEN_STYLESHEET = ".root {-fx-base: green}";
+    private static final String FX_BASE_GRAY_STYLESHEET = ".root {-fx-base: #808080}";
+
+    /**
+     * All cases will lead to a neutral gray color #808080
+     */
+    enum OverrideCases {
+        // User agent styles win when not overridden directly or indirectly:
+        USER_AGENT(GRAY_STYLESHEET, null, null, null),
+        INDIRECT_USER_AGENT(GRAY_INDIRECT_STYLESHEET, null, null, null),
+
+        // Property wins when not directly overridden by author or inline style:
+        PROPERTY_OVERRIDES_USER_AGENT(RED_STYLESHEET, Color.web("#808080"), null, null),
+        PROPERTY_OVERRIDES_INDIRECT_USER_AGENT(RED_INDIRECT_STYLESHEET, Color.web("#808080"), null, null),
+        PROPERTY_OVERRIDES_USER_AGENT_VARIABLE_SET_IN_AUTHOR(RED_INDIRECT_STYLESHEET, Color.web("#808080"), FX_BASE_GREEN_STYLESHEET, null),
+        PROPERTY_OVERRIDES_USER_AGENT_VARIABLE_SET_INLINE(RED_INDIRECT_STYLESHEET, Color.web("#808080"), null, "-fx-base: yellow"),
+
+        // Author style wins when not directly overridden by inline style:
+        AUTHOR_OVERRIDES_USER_AGENT(RED_STYLESHEET, null, GRAY_STYLESHEET, null),
+        AUTHOR_OVERRIDES_INDIRECT_USER_AGENT(RED_INDIRECT_STYLESHEET, null, GRAY_STYLESHEET, null),
+        AUTHOR_OVERRIDES_PROPERTY(RED_STYLESHEET, Color.BLUE, GRAY_STYLESHEET, null),
+        AUTHOR_OVERRIDES_USER_AGENT_VARIABLE_SET_INLINE(RED_INDIRECT_STYLESHEET, null, GRAY_STYLESHEET, "-fx-base: yellow"),
+
+        // Indirect author styles win when property is not set directly, and there is no direct or indirect override in an inline style:
+        AUTHOR_VARIABLE_OVERRIDES_USER_AGENT_VARIABLE(RED_INDIRECT_STYLESHEET, null, FX_BASE_GRAY_STYLESHEET, null),
+
+        // Direct inline styles always win:
+        INLINE_OVERRIDES_USER_AGENT(RED_STYLESHEET, null, null, "-fx-background-color: #808080"),
+        INLINE_OVERRIDES_PROPERTY(RED_STYLESHEET, Color.BLUE, null, "-fx-background-color: #808080"),
+        INLINE_OVERRIDES_AUTHOR(RED_STYLESHEET, null, RED_STYLESHEET, "-fx-background-color: #808080"),
+
+        // Indirect inline styles win when not directly overridden by author stylesheet or property:
+        INLINE_VARIABLE_OVERRIDES_USER_AGENT_VARIABLE(RED_INDIRECT_STYLESHEET, null, null, "-fx-base: #808080"),
+        INLINE_VARIABLE_OVERRIDES_AUTHOR_VARIABLE(RED_INDIRECT_STYLESHEET, null, FX_BASE_GREEN_STYLESHEET, "-fx-base: #808080");
+
```

Comment:
> Another question: do we want to go through all the cases of Scene/SubScene/Parent/Region user agent style sheet, for completeness sake, or would that be a different test file?

In my opinion, this does seem a bit of out of scope for this change.  it would for example not add any new coverage to the code that was changed.  Also the origin of the user agent stylesheet is not a factor in the resolution of lookups.  The user agent stylesheets, their replacement (via Scene/Subscene) or addition (via Region) should (already?) be tested elsewhere.

Reply:
I added a few cases where user agent is `null`.  I've not added all combinations as that would lead to 20-30 extra cases that add little of value.  If I already verified that INLINE will override the lower priority sources individually, then it's safe to say that all possible other combinations of those lower priority sources will also get overridden.

Reply:
I still think there is a value in providing exhaustive coverage, especially since the number of combination is relatively low.  The added value, for instance, is to guard against regressions introduced by other people.

Reply:
@andy-goryachev-oracle 

Could you be more specific which cases (related to this change and lookups) are currently not being adequately covered?
  
I'm of the opinion that the tests are more than sufficient for the changes involved, I don't think it is the responsibility of this PR to add coverage in areas unrelated to it.  The cases you are requesting IMHO check cascading rules, which should already be verified elsewhere (and if not, is certainly not something that should be part of this PR).

The tests I added include 19(!) new cases (versus the 13 that existed before), but don't add **any** new coverage (lookups were already covered by for example `removingThenAddingNodeToDifferentBranchGetsIneritableStyle`) -- they're only adding assertions specifically for this change.  One might even say I overdid it here, as half of the tests don't even deal with lookups (ie. all the cases where no "indirect" variant is involved). Making these test cases exhaustive for something that has 4 variables and 3-4 options each would result in 80+ cases.

So adding further tests does not seem to serve any purpose from where I stand; most of the requested additional cases would not even involve lookups, or would just confirm existing cascading rules (if an inline style overrides a property **or** author style, then it would also override property **and** author style). And none of them would extra coverage...

Reply:
fine